internal/proxy/reverse_proxy.go

```diff
@@ -0,0 +1,161 @@
+	"net/http/httputil"
+	"net/url"
+	"strings"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+)
+
+const (
+	OpenshiftGatewayTimeout = 5 * time.Second
+)
+
+type reverseProxy struct {
+	redirectURL     url.URL
+	logger          *log.Entry
+	responseTimeout time.Duration
+}
+
+type directorFunc func(req *http.Request)
+
+func director(target *url.URL) directorFunc {
+	// targetQuery := target.RawQuery
+	return func(req *http.Request) {
+		log.Warnf(" > RevProxy: CALLING Director  | target:  %v | req: %q ", target, req.URL)
+		if _, ok := req.Header["User-Agent"]; !ok {
+			// explicitly disable User-Agent so it's not set to default value
+			req.Header.Set("User-Agent", "")
+		}
+		log.Infof(" > RevProxy: Director %v", req.URL)
+	}
+
+}
+
+func newLoggedReverseProxy(target *url.URL) *httputil.ReverseProxy {
+	// copied from:
+
+	transport := &http.Transport{
+		Proxy: func(req *http.Request) (*url.URL, error) {
+			// log.Println(" > RevProxy: CALLING Proxy")
+			return http.ProxyFromEnvironment(req)
+		},
+
+		Dial: func(network, addr string) (net.Conn, error) {
+			// log.Println(" > RevProxy: CALLING Dial")
+
+			dialer := &net.Dialer{
+				Timeout:   30 * time.Second,
+				KeepAlive: 30 * time.Second,
+			}
+
+			conn, err := dialer.Dial(network, addr)
+			if err != nil {
+				log.Infof(" > RevProxy: Dial - Error: %v", err)
+			}
+			return conn, err
+		},
+
+		TLSHandshakeTimeout: 10 * time.Second,
+	}
+
+	return &httputil.ReverseProxy{
+		Director:  director(target),
+		Transport: transport,
+		// ModifyResponse: modifyResponse,
+	}
+}
+
+type responseRecorder struct {
+	http.ResponseWriter
+	statusCode  int
+	err         error
+	wroteHeader bool
+	log         *log.Entry
+}
+
+func newResponseRecorder(w http.ResponseWriter, logger *log.Entry) *responseRecorder {
+	return &responseRecorder{
+		ResponseWriter: w,
+		log:            logger.WithField("response_recorder", "true"),
+		statusCode:     http.StatusOK,
+		err:            nil,
+		wroteHeader:    false,
+	}
+}
+
+func (rr *responseRecorder) WriteHeader(code int) {
+	rr.log.Infof("Write Header called with %d", code)
+
+	if rr.wroteHeader {
+		rr.log.Warnf("Multiple Write Header %d called; SKIPPED", code)
+		return
+	}
+
+	rr.statusCode = code
+	rr.wroteHeader = true
+
+	// refer httputils/reverseproxy.go ServerHTTP which sets the
+	// StatusBadGateway when transport.RoundTrip returns an error
+	if code == http.StatusBadGateway {
+		// set err for bad gateway
+		rr.err = fmt.Errorf("bad gateway error: %d", code)
+		rr.log.Warnf("Error BadGateway WriteHeader %v - SKIPPED", code)
+		return
+	}
+
+	rr.log.Infof(" > RevProxy: WriteHeader %v - OK", code)
+	rr.ResponseWriter.WriteHeader(code)
+}
+
+func (rr *responseRecorder) Write(stuff []byte) (int, error) {
+	if rr.err == nil {
+		rr.log.Infof(" > RevProxy: Write stuff len: %d - OK", len(stuff))
+		return rr.ResponseWriter.Write(stuff)
+	}
+
+	// fake that it wrote it
+	rr.log.Warnf("Faking write len: %d - SKIPPED as err is set", len(stuff))
```

Comment:
`Write` is supposed to return the length of the content it wrote, but if there has been an `err` before then the response isn't usable as we should send a `302` so instead of writing the data we just return that we wrote the data without any issue.

Reply:
just noticed, the `stuff` must be called `data`

Reply:
@kishansagathiya 

> LGTM. I wonder if we should move responseRecorder to a different file.


Yes, I will split the reverseproxy part to 
- `proxy/reverseproxy/reverseproxy.go`
- `proxy/reverseproxy/reverseproxy_test.go`
- `proxy/reverseproxy/response.go` 
- `proxy/reverseproxy/response_test.go`